src/app.rs

```diff
@@ -71,16 +71,38 @@ where
         InitError = (),
     >,
 {
-    /// Set application data. Application data could be accessed
-    /// by using `Data<T>` extractor where `T` is data type.
+    /// Adds an arbitrary application level data item.
     ///
-    /// **Note**: http server accepts an application factory rather than
-    /// an application instance. Http server constructs an application
-    /// instance for each thread, thus application data must be constructed
-    /// multiple times. If you want to share data between different
-    /// threads, a shared object should be used, e.g. `Arc`. Internally `Data` type
-    /// uses `Arc` so data could be created outside of app factory and clones could
-    /// be stored via `App::app_data()` method.
+    /// Application data can be accessed by using a `Data<T>` extractor where `T` is the data type.
+    ///
+    /// The state is managed on a per-type basis and as such there can be
+    /// at most one value for any given type.
+    /// This means that only the first invocation of this function per type will have an effect,
+    /// all later ones will be ignored.
```

Comment:
Is this true, or is it the last invocation is used.

Reply:
I did look at the code back then and I thought this is what it does but I can take another look.

Reply:
I have to be honest: I don't understand (anymore) how this is used. If one of you has any insights I'll happily add it otherwise I'd just remove that sentence.

It would be good to document the behavior though.

Reply:
You know what you're right, and that's surprising behavior actually because .app_data works differently; a bug, even.

Reply:
Out of curiosity (because I couldn't trace it anymore but I know I did in the past): Can you point me at the location where this happens?